### PR TITLE
[UI] Volume Batch Creation

### LINKED ISSUE #2964: Batch Volume Creation from MetalK8s UI
**Component**: ui, volumes

**Why this is needed**:
We want to be able to batch create volume from UI.

**What should be done**:
image
image
Please check the HD from Sketch:

https://www.sketch.com/s/b7c9bece-ad22-40e6-b2b5-410645289650/a/qP32vd
https://www.sketch.com/s/b7c9bece-ad22-40e6-b2b5-410645289650/a/L5Vox5

**Implementation proposal** (strongly recommended):
When we choose `SparseLoop` Device, we can have the size field for each of the volume instead of path.

The number selection component is missing in core-ui. I would suggest first implement in MetalK8s and generalize it to core-ui.

For the moment, I only see `Device Path` require question mark next to it with the following text as tooltip.
`A devicePath or a partition mount point. Example: /dev/sdb`

We apply the same logic as the Scality Cloud does. 
when selecting the multi-volume creation, by default we should have 1 volume to create.

We should dispatch the data in an array. 
```js
  [
    { name: "volume01'", devicePath: '/dev/sdb' },
    { name: "volume02'", devicePath: '/dev/sdc' },
  ];
```
**Test plan**:
We should implement the interaction test for the batch volume creation.

### PR DESCRIPTION
**Pull request by ChengYanJin**
**Component**: ui, volumes 

**Context**: 
We want to be able to batch create volume from UI.

**Summary**:
Note that:
For `RawBlockDevice`, the recommended path only happens with `'/dev/vda'`, we will see later if 
For `SparseLoopDevice`, the size of the batch volume are the same as the default one.

**Acceptance criteria**: 
1. The Default Volume Creation Form
image

2. Volume type `RawBlockDevice` batch creation 
image

3. Volume type `SparseLoopDevice` batch creation. (From big HD screen)
image

---

Closes: #2964

### CODE REVIEW

File: ui/src/services/utils.js [hunk 1]
```diff
@@ -419,3 +419,45 @@ export const useDynamicChartSize = (container_id) => {
...
+ *
+ * @example
+ * const string = '/dev/vda'
+ *
+ * const nextDevicePath = linuxDrivesNamingIncrement(string)
+ */
+export const linuxDrivesNamingIncrement = (string, increment) => {
+  if (string.match(/^\/dev\/vd[a-z]/)) {
+    while (increment--) {
+      var lastChar = string.slice(-1);
```
Comment (JBWatenbergScality) [inline ui/src/services/utils.js]: I'm not yet sure about your styleguide conventions but I would always prefer usage of let or const for variable definitions, using const here will reflect clearly that this variable doesn't change elsewhere and will slightly simplify reading this code

File: ui/src/services/utils.js [hunk 2]
```diff
@@ -419,3 +419,45 @@ export const useDynamicChartSize = (container_id) => {
...
+ * 18278	    vdzzz
+ *
+ * @param {string} string - The string that will be incremented
+ *
+ * @example
+ * const string = '/dev/vda'
+ *
+ * const nextDevicePath = linuxDrivesNamingIncrement(string)
+ */
+export const linuxDrivesNamingIncrement = (string, increment) => {
```
Comment (JBWatenbergScality) [inline ui/src/services/utils.js]: I would rename `string` to something more explicit such as `drivePath`

File: ui/src/services/utils.js [hunk 3]
```diff
@@ -419,3 +419,45 @@ export const useDynamicChartSize = (container_id) => {
...
+ * 703	      vdaaa
+ * 704	      vdaab
+ * 18278	    vdzzz
+ *
+ * @param {string} string - The string that will be incremented
+ *
+ * @example
+ * const string = '/dev/vda'
+ *
+ * const nextDevicePath = linuxDrivesNamingIncrement(string)
```
Comment (JBWatenbergScality) [inline ui/src/services/utils.js]: This example lack increment parameter

File: ui/src/services/utils.js (hunk 2 above)
Comment (JBWatenbergScality) [reply ui/src/services/utils.js]: Additionally for utils function like this one it seems important for me to add a unit test that covers the edge cases you want to support, but not sure if it is your traditional way of working so we can discuss about it.

File: ui/src/services/utils.js [hunk 4]
```diff
@@ -419,3 +419,45 @@ export const useDynamicChartSize = (container_id) => {
 
   return [graphWidth, window.innerHeight / 6 - 30];
 };
+
+/**
+ * This function increments the string by 1
```
Comment (JBWatenbergScality) [inline ui/src/services/utils.js]: This description doesn't seems to be accurate

File: ui/src/containers/CreateVolume.js [hunk 5]
```diff
@@ -237,16 +241,55 @@ const CreateVolume = (props) => {
...
+      }
+    }, [path, touched.path, setFieldValue, props.name, props.index]);
+
+    return (
+      <>
+        <Input {...props} {...field} />
+        {!!meta.touched && !!meta.error && <div>{meta.error}</div>}
+      </>
+    );
   };
```
Comment (JBWatenbergScality) [inline ui/src/containers/CreateVolume.js]: As those two components only differs in the way they setFieldValue, they might be factorized.

File: ui/src/containers/CreateVolume.js [hunk 6]
```diff
@@ -362,6 +413,34 @@ const CreateVolume = (props) => {
...
 
+              // Update the volume list base on the current input number
+              const setVolumeNumber = (field, arrayHelpers) => (
+                selectedObj,
+              ) => {
+                const inputVolNum = selectedObj.target.value;
+                const preVolNum = values.numberOfVolumes;
+
+                setFieldValue(field, inputVolNum);
+                var diff = preVolNum - inputVolNum;
```
Comment (JBWatenbergScality) [inline ui/src/containers/CreateVolume.js]: I would use const instead to declare this variable

File: ui/src/containers/CreateVolume.js [hunk 7]
```diff
@@ -362,6 +413,34 @@ const CreateVolume = (props) => {
...
+                if (diff > 0) {
+                  // REMOVE volume object from `values.volumes` base on the index
+                  for (var i = inputVolNum; i < preVolNum; i++) {
```
Comment (JBWatenbergScality) [inline ui/src/containers/CreateVolume.js]: I would prefer use let to avoid any hoisting issues

File: ui/src/containers/CreateVolume.js [hunk 8]
```diff
@@ -362,6 +413,34 @@ const CreateVolume = (props) => {
...
+                    arrayHelpers.remove(i);
+                  }
+                } else if (diff < 0) {
+                  // PUSH new volume object to `values.volumes`
+                  var absDiff = Math.abs(diff);
```
Comment (JBWatenbergScality) [inline ui/src/containers/CreateVolume.js]: I would use const instead to declare this variable

File: ui/src/ducks/app/volumes.js [hunk 9]
```diff
@@ -118,8 +118,8 @@ export const updateStorageClassAction = (payload) => {
   return { type: UPDATE_STORAGECLASS, payload };
 };
 
-export const createVolumeAction = (newVolume) => {
-  return { type: CREATE_VOLUMES, payload: { newVolume } };
+export const createVolumeAction = (newVolumes) => {
```
Comment (JBWatenbergScality) [inline ui/src/ducks/app/volumes.js]: I would rename this to `createVolumesAction` to reflect the batch creation possibility

File: ui/src/ducks/app/volumes.js [hunk 10]
```diff
@@ -205,97 +205,97 @@ export function* fetchStorageClass() {
...
+      }
     } else {
+      // We might want to change this behavior later
       yield put(
         addNotificationErrorAction({
-          title: intl.translate('volume_creation'),
-          message: intl.translate('volume_creation_failed', {
-            name: newVolume.name,
-          }),
+          title: 'Volume Form Error',
```
Comment (JBWatenbergScality) [inline ui/src/ducks/app/volumes.js:294]: Translation seems to be missing here

File: ui/src/containers/CreateVolume.js [hunk 11]
```diff
@@ -54,6 +55,8 @@ const ActionContainer = styled.div`
 const CreateVolumeLayout = styled.div`
   display: inline-block;
   margin-top: ${padding.base};
+  overflow-y: scroll;
```
Comment (JBWatenbergScality) [inline ui/src/containers/CreateVolume.js]: Is it intended to display scrollbars by default even for non scrollable content ? You might prefer to use `auto` instead

File: ui/src/containers/CreateVolume.js [hunk 12]
```diff
@@ -622,7 +640,22 @@ const CreateVolume = (props) => {
...
                         onChange={handleChange('path')}
-                        label={intl.translate('device_path')}
+                        label={
+                          <>
+                            {intl.translate('device_path')}
+                            <Tooltip
+                              placement="right"
+                              overlay={
+                                <div style={{ minWidth: '200px' }}>
+                                  A devicePath or a partition mount point.
```
Comment (JBWatenbergScality) [inline ui/src/containers/CreateVolume.js]: I think there is a missing translation here

File: ui/src/containers/CreateVolume.js [hunk 13]
```diff
@@ -290,9 +382,18 @@ const CreateVolume = (props) => {
             initialValues={initialValues}
             validationSchema={validationSchema}
             onSubmit={(values) => {
-              const newVolume = { ...values };
-              newVolume.size = `${values.sizeInput}${values.selectedUnit}`;
-              createVolume(newVolume);
+              const newVolumes = { ...values };
+              newVolumes.size = `${values.sizeInput}${values.selectedUnit}`;
+              if (newVolumes.multiVolumeCreation) {
```
Comment (JBWatenbergScality) [inline ui/src/containers/CreateVolume.js]: This if could be transferred in formatDataForBatchVolumeCreation as it already partially exists there and just missing the else branch to return a one element array

File: ui/src/services/NodeVolumesUtils.js [hunk 14]
```diff
@@ -276,3 +276,28 @@ export const getVolumeListData = createSelector(
     return nodeVolumes.sort((a, b) => compareHealth(b.health, a.health));
   },
 );
+
+export const formatDataForBatchVolumeCreation = (newVolumes) => {
```
Comment (JBWatenbergScality) [inline ui/src/services/NodeVolumesUtils.js]: I would add a simple unit test for the two branches of this function, also as mentioned in the usage of this function I would handle the `!multiVolumeCreation` case here and return a single entry array containing newVolumes, that would allow a simpler usage of the function as its ouput will always be of the same shape `(newVolumes: VolumeCreationFormState) => Volume[]` instead of `(newVolumes: VolumeCreationFormState) => Volume[] | void`

File: ui/src/services/utils.js (hunk 2 above)
Comment (ChengYanJin) [reply ui/src/services/utils.js]: Indeed, it's a very bad name. I will rename it to `devicePath`.

**Review by thomasdanan — DISMISSED** [review]
@ChengYanJin  I like the behavior.
I would change one little thing:
When you create multiple Volumes, the Device Path for each Volume is computed from the Global Device Path field. I think the first Volume should get what you input in the Global Device Path field (i.e. not increment the Device Path of the first Volume).

Also when playing with multiple Volume check box and the number of Volumes, you will observe strange behavior. Maybe some state not properly managed

I wonder if we should disable Global Name and Device Path global fields when creating multiple Volumes because if you override Name or Device Path and then you change the Global field, you will lose all your customizations (@JBWatenbergScality what do you think?)

Also for number increments, I would do 01, 02, 03 etc ... instead of 1, 2, 3 .... 

Thanks and well done

**Comment by JBWatenbergScality** [comment]
@thomasdanan @ChengYanJin From my point of view once the name or path is touched we shouldn't change automatically their values to maintain the user intentions but should still let the ability for the user to change the first volume name and path. 

Also the ui might be a little bit misleading as the user is unsure if the default values will actually create a volume. To be precise mentioning `/dev/vda` by default and then creating one more volume will pre-complete the first entry with `/dev/vdb` in the current behavior but when submitting the form just `/dev/vdb` will be created.

The simplest interaction I could think of closed enough to the current implementation would be to remove the default name and path fields when checking `Create multiple volumes?` and compute the names of the volumes from the first one if subsequent one are not touched.

File: ui/src/services/utils.js [hunk 15]
```diff
@@ -479,3 +479,23 @@ export const linuxDrivesNamingIncrement = (devicePath, increment) => {
     return '';
   }
 };
+
+/**
+ * This function formats the name base on the index
+ * @param {string} name - The name the default volume name
```
Comment (JBWatenbergScality) [inline ui/src/services/utils.js]: Remove redundant name `The name the default volume`

File: ui/src/services/utils.js [hunk 16]
```diff
@@ -479,3 +479,23 @@ export const linuxDrivesNamingIncrement = (devicePath, increment) => {
...
+ * @param {number} index  - The number of index
+ *
+ * @example
+ * const name = 'volume-test'
+ *
+ * const nextDevicePath = linuxDrivesNamingIncrement(name, 1)
```
Comment (JBWatenbergScality) [inline ui/src/services/utils.js]: `const formatedName = formatBatchName(name, 1) `

File: ui/src/containers/CreateVolume.js [hunk 17]
```diff
@@ -513,12 +633,108 @@ const CreateVolume = (props) => {
...
+                            >
+                              <span
+                                style={{
+                                  paddingRight: `${padding.base}`,
+                                }}
+                              >
+                                {intl.translate('number_volume_create')}
+                              </span>
+                              {/* TODO: Generalize the number input component to core-ui. */}
+                              <InputNumberComponentStyle
```
Comment (JBWatenbergScality) [inline ui/src/containers/CreateVolume.js:711]: I think we should handle a max number of volumes to avoid dom freeze in case of the user mistakenly ask 10000 volumes, simply add a max prop here should be enough

Comment (ChengYanJin) [reply ui/src/containers/CreateVolume.js:711]: Indeed, good point!
I am going to set the `max` number to`70` for the moment.

**Comment by ChengYanJin** [comment]
Hello @thomasdanan @JBWatenbergScality ,

According to the feedback from you and our offline discussions. I've addressed the following issues. 

1.
> Thomas: When you create multiple Volumes, the Device Path for each Volume is computed from the Global Device Path field. I think the first Volume should get what you input in the Global Device Path field (i.e. not increment the Device Path of the first Volume).

Fix the bug of the Device path, the first volume should be the same as the Global field.

2.
> Thomas: Also when playing with multiple Volume check box and the number of Volumes, you will observe strange behavior. Maybe some state not properly managed

Fix the bug, will add interaction test of the form in the following PR.

3.
> Thomas: I wonder if we should disable Global Name and Device Path global fields when creating multiple Volumes because if you override Name or Device Path and then you change the Global field, you will lose all your customizations (@JBWatenbergScality what do you think?)

 In order to keep the user customization, edit the global value(name/path) should not affect the defaults pre-fill of the touched field.
Unless the users uncheck the box of multi-volume creation and re-check again, we only update the untouched field with the new global value.

4. 
> Thomas: Also for number increments, I would do 01, 02, 03 etc ... instead of 1, 2, 3 ....

Done.

5. Set the maximum number of the volume to create `70` to prevent the UI frozen. 
The UI should pop up the error message when the input number is bigger than 70 (Waiting for the [PR #213](https://github.com/scality/core-ui/pull/213) in core-ui to be landed)

Please let me know if you see any improvements/ bugs to be done. 
Thank you!

File: ui/src/containers/CreateVolume.js [hunk 18]
```diff
@@ -421,6 +421,33 @@ const CreateVolume = (props) => {
...
+                  }
+                  if (
+                    touched &&
+                    touched.volumes &&
+                    touched?.volumes[i]?.path === true
+                  ) {
+                    continue;
+                  } else {
+                    setFieldValue(`volumes[${i}]path`, '');
+                  }
```
Comment (JBWatenbergScality) [inline ui/src/containers/CreateVolume.js]: I think this can be simplified to : 
```js
if (!touched?.volumes[i]?.name) {
    setFieldValue(`volumes[${i}]name`, '');
}
if (!touched?.volumes[i]?.path) {
    setFieldValue(`volumes[${i}]path`, '');
}
```

Comment (ChengYanJin) [reply ui/src/containers/CreateVolume.js]: Hi @JBWatenbergScality ,

Yes. Good point. But I guess we need to use optional chaining to make sure the Volumes array exists.
Please see:

```js
                  if (!touched?.volumes?.[i]?.name) {
                    setFieldValue(`volumes[${i}]name`, '');
                  }
                  if (!touched?.volumes?.[i]?.path) {
                    setFieldValue(`volumes[${i}]path`, '');
                  }
```

**Review by JBWatenbergScality — APPROVED** [review]
